### PR TITLE
FIX constrained_layout w/ hidden axes

### LINKED ISSUE #14918: constrained_layout fails with hidden axis...
### Bug report

```python
import numpy as np
import matplotlib.pyplot as plt

fig5, axs = plt.subplots(2, 2, constrained_layout=True)

axs[0,1].set_visible(False)

plt.show()
```

will yield an error:

```
Traceback (most recent call last):
  File "/Users/jklymak/matplotlib/lib/matplotlib/backends/backend_qt5.py", line 488, in _draw_idle
    self.draw()
  File "/Users/jklymak/matplotlib/lib/matplotlib/backends/backend_agg.py", line 396, in draw
    self.figure.draw(self.renderer)
  File "/Users/jklymak/matplotlib/lib/matplotlib/artist.py", line 38, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/Users/jklymak/matplotlib/lib/matplotlib/figure.py", line 1711, in draw
    self.execute_constrained_layout(renderer)
  File "/Users/jklymak/matplotlib/lib/matplotlib/figure.py", line 2433, in execute_constrained_layout
    do_constrained_layout(fig, renderer, h_pad, w_pad, hspace, wspace)
  File "/Users/jklymak/matplotlib/lib/matplotlib/_constrained_layout.py", line 170, in do_constrained_layout
    _make_layout_margins(ax, renderer, h_pad, w_pad)
  File "/Users/jklymak/matplotlib/lib/matplotlib/_constrained_layout.py", line 275, in _make_layout_margins
    bbox = invTransFig(tightbbox)
  File "/Users/jklymak/matplotlib/lib/matplotlib/transforms.py", line 1452, in transform_bbox
    return Bbox(self.transform(bbox.get_points()))
AttributeError: 'NoneType' object has no attribute 'get_points'
```

**Matplotlib version**

  * Matplotlib version: Master

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

If an axes is hidden (set_visible(False)) then it has no tightbbox.  But we still need to position it w/ `constrained_layout`, pointed out in #14917 (thanks @ImportanceOfBeingErnest )

Closes #14918

```python
import numpy as np
import matplotlib.pyplot as plt

fig5, axs = plt.subplots(2, 2, constrained_layout=True)
axs[0,1].set_visible(False)
plt.show()
```

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
Could easily be back ported if not a bother.

**Comment by ImportanceOfBeingErnest** [comment]
Should a non-visible axes take part in the layout? If the answer to that is yes, this is a good fix; but one might also consider that a non-visible axes may just be treated as "not there", or maybe as the most tiny box possible at the center of its original position?

**Comment by jklymak** [comment]
An empty gridspec box gets a placeholder like this anyhow, so this is the smallest box that is possible for this spot if gridspec or `subplots` has been used for the layout.  

I do this because if someone has a gap in their layout presumably they want it to be there.  And it is needed to keep the axes in rows above or below a blank (for instance) aligned.  

This is somewhat poorly documented at: https://matplotlib.org/3.1.1/tutorials/intermediate/constrainedlayout_guide.html#empty-gridspec-slots
and its checked in one of the tests... (the horribly named: `test_constrained_layout8`)

**Comment by ImportanceOfBeingErnest** [comment]
Yes, I realized that 5 minutes after I posted (see edit above), sorry for the confusion.

**Comment by jklymak** [comment]
Too fast!  Though I'll point out that even remove will still put a blank gridspec in there (actually an axes with all the elements made invisible!)

File: lib/matplotlib/_constrained_layout.py [hunk 1]
```diff
@@ -272,7 +272,11 @@ def _make_layout_margins(ax, renderer, h_pad, w_pad):
     invTransFig = fig.transFigure.inverted().transform_bbox
     pos = ax.get_position(original=True)
     tightbbox = ax.get_tightbbox(renderer=renderer)
-    bbox = invTransFig(tightbbox)
+    if tightbbox is None:
```
Comment (timhoffm) [inline lib/matplotlib/_constrained_layout.py]: Just:
~~~
bbox = pos if tightbbox is None else invTransFig(tightbbox)
~~~

Comment (jklymak) [reply lib/matplotlib/_constrained_layout.py]: I find those really hard to read a lot of the time.

File: lib/matplotlib/tests/test_constrainedlayout.py [hunk 2]
```diff
@@ -387,3 +387,13 @@ def test_colorbar_location():
     fig.colorbar(pcm, ax=axs[-2, 3:], shrink=0.5, location='top')
     fig.colorbar(pcm, ax=axs[0, 0], shrink=0.5, location='left')
     fig.colorbar(pcm, ax=axs[1:3, 2], shrink=0.5, location='right')
+
+
+def test_hidden_axes():
```
Comment (dstansby) [inline lib/matplotlib/tests/test_constrainedlayout.py]: Can you add a quick comment here with a mention of this issue?

Comment (jklymak) [reply lib/matplotlib/tests/test_constrainedlayout.py]: Added: 
```
    # test that if we make an axes not vidsdible that constrained_layout
    # still works.  Note the axes still takes space in the layout
    # (as does a gridspect slot that is empty)
```

File: lib/matplotlib/tests/test_constrainedlayout.py [hunk 3]
```diff
@@ -387,3 +387,16 @@ def test_colorbar_location():
...
+    # test that if we make an axes not vidsdible that constrained_layout
```
Comment (dstansby) [inline lib/matplotlib/tests/test_constrainedlayout.py]:
```suggestion
    # test that if we make an axes not visible that constrained_layout
```

File: lib/matplotlib/tests/test_constrainedlayout.py [hunk 4]
```diff
@@ -387,3 +387,16 @@ def test_colorbar_location():
...
+    # still works.  Note the axes still takes space in the layout
+    # (as does a gridspect slot that is empty)
```
Comment (dstansby) [inline lib/matplotlib/tests/test_constrainedlayout.py]:
```suggestion
    # (as does a gridspec slot that is empty)
```

Comment (jklymak) [reply lib/matplotlib/tests/test_constrainedlayout.py]: I have one of the accursed mac keyboards... ;-)

**Comment by timhoffm** [comment]
flake8:

> ./lib/matplotlib/_constrained_layout.py:279:1: W293 blank line contains whitespace

**Comment by jklymak** [comment]
Hmmm, I kept trying to rebase that change in, but I suspect I added the space in the "TST" commit, so it kept overwriting it.  Ooops....

File: lib/matplotlib/_constrained_layout.py (hunk 1 above)
Comment (timhoffm) [reply lib/matplotlib/_constrained_layout.py]: Actually, when reading a function for a basic understanding what it does I find this more easy.

It's just one line saying: "Sets `bbox` depending on some condition". The if/else is already more low level.

But not fighting over this.

**Pull request merged**